core/ledger/kvledger/txmgmt/privacyenabledstate/snapshot.go

```diff
@@ -0,0 +1,184 @@
+
+// ExportPubStateAndPvtStateHashes generates four files in the specified dir. The files, public_state.data and public_state.metadata
+// contains the exported public state and the files private_state_hashes.data and private_state_hashes.data contain the exported private state hashes.
+// The file format for public state and the private state hashes are the same. The data files contains a series of tuple <key,value> and the metadata
+// files contains a series of tuple <namespace, num entries for the namespace in the data file>.
+func (s *DB) ExportPubStateAndPvtStateHashes(dir string, newHasher func() hash.Hash) (map[string][]byte, error) {
+	itr, dbValueFormat, err := s.GetFullScanIterator(isPvtdataNs)
+	if err != nil {
+		return nil, err
+	}
+	defer itr.Close()
+
+	pubStateWriter, err := newSnapshotWriter(
+		path.Join(dir, pubStateDataFileName),
+		path.Join(dir, pubStateMetadataFileName),
+		dbValueFormat,
+		newHasher,
+	)
+	if err != nil {
+		return nil, err
+	}
+	defer pubStateWriter.close()
+
+	pvtStateHashesWriter, err := newSnapshotWriter(
+		path.Join(dir, pvtStateHashesFileName),
+		path.Join(dir, pvtStateHashesMetadataFileName),
+		dbValueFormat,
+		newHasher,
+	)
+	if err != nil {
+		return nil, err
+	}
+	defer pvtStateHashesWriter.close()
+
+	for {
+		compositeKey, dbValue, err := itr.Next()
+		if err != nil {
+			return nil, err
+		}
+		if compositeKey == nil {
+			break
+		}
+		switch {
+		case isHashedDataNs(compositeKey.Namespace):
+			if err := pvtStateHashesWriter.addData(compositeKey, dbValue); err != nil {
+				return nil, err
+			}
+		default:
+			if err := pubStateWriter.addData(compositeKey, dbValue); err != nil {
+				return nil, err
+			}
+		}
+	}
+	pubStateDataHash, pubStateMetadataHash, err := pubStateWriter.done()
+	if err != nil {
+		return nil, err
+	}
+	pvtStateHahshesDataHash, pvtStateHashesMetadataHash, err := pvtStateHashesWriter.done()
+	if err != nil {
+		return nil, err
+	}
+	return map[string][]byte{
+			pubStateDataFileName:           pubStateDataHash,
+			pubStateMetadataFileName:       pubStateMetadataHash,
+			pvtStateHashesFileName:         pvtStateHahshesDataHash,
+			pvtStateHashesMetadataFileName: pvtStateHashesMetadataHash,
+		},
+		nil
+}
+
+// snapshotWriter generates two files, a data file and a metadata file. The datafile contains a series of tuples <key, dbValue>
+// and the metadata file contains a series of tuples <namesapce, number-of-tuples-in-the-data-file-that-belong-to-this-namespace>
+type snapshotWriter struct {
+	dataFile                *snapshot.FileWriter
+	metadataFile            *snapshot.FileWriter
+	kvCountsPerNamespace    map[string]uint64
+	namespaceInsertionOrder []string
+}
+
+func newSnapshotWriter(
+	dataFilePath, metadataFilePath string,
+	dbValueFormat byte,
+	newHasher func() hash.Hash,
+) (*snapshotWriter, error) {
+
+	var dataFile, metadataFile *snapshot.FileWriter
+	var err error
+	defer func() {
+		if err != nil {
+			dataFile.Close()
+			metadataFile.Close()
```

Comment:
method invocation on nil pointer. As the Close() method is aware of this, no issue here. Found the following article from golang experience report (https://github.com/golang/go/wiki/ExperienceReports) -- [Fixing the billion-dollar mistake in Go by borrowing from Rust
 ](https://getstream.io/blog/fixing-the-billion-dollar-mistake-in-go-by-borrowing-from-rust/).